core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,318 @@
+
+        // index of the leaf path in map. We store paths of neighboring leaves in one key-value pair
+        let half_pos = (pos / 2) as u64;
+
+        // Now we returning an error, if the new path is not the same as the one in the map
+        let mut extended_path = path;
+        if is_even(pos) {
+            extended_path.insert(0, value);
+        } else {
+            extended_path.insert(1, value);
+        }
+
+        let root_of_current_path = compute_root(&extended_path, depth, index);
+        if self.root == [Felt::ZERO; 4] {
+            self.root = root_of_current_path;
+        } else if self.root != root_of_current_path {
+            return Err(AdviceSetError::InvalidPath(extended_path));
+        }
+        self.paths.insert(half_pos, extended_path);
+
+        Ok(())
+    }
+
+    /// Returns the root of parent Merkle tree.
+    pub fn root(&self) -> Word {
+        self.root
+    }
+
+    /// Returns the depth of this Merkle tree.
+    ///
+    /// Merkle tree of depth 1 has two leaves, depth 2 has four leaves etc.
+    pub fn depth(&self) -> u32 {
+        self.total_depth
+    }
+
+    /// Returns a node at the specified index.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    pub fn get_node(&self, depth: u32, index: u64) -> Result<Word, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = (pos / 2) as u64;
+        if is_even(pos) {
+            Ok(self.paths[&index][0])
+        } else {
+            Ok(self.paths[&index][1])
+        }
+    }
+
+    /// Returns a Merkle path to the node at the specified index. The node itself is
+    /// not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    pub fn get_path(&self, depth: u32, index: u64) -> Result<Vec<Word>, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = pos / 2;
+
+        let mut path = self.paths[&index].clone();
+
+        if is_even(pos) {
+            path.remove(0);
+            Ok(path)
+        } else {
+            path.remove(1);
+            Ok(path)
+        }
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if the node with specified index is not in the map.
```

Comment:
Similar to the above comments: we should probably return an error if the path does not exist in this set.